src/core/utils/EventNormalizer.ts

```diff
@@ -222,79 +265,244 @@ export class EventNormalizer {
+        // observed events.
+        compiledEvents.forEach(compiledEvent => {
+            const events = this._analyze(compiledEvent, mutatedElements);
+            customEvents.push(...events);
+        });
+
+        this._triggerEvent({ events: customEvents, mutatedElements });
+    }
+    /**
+     * @private
+     * @param {KeyboardEvent} ev
+     * @returns CompiledEvent
+     */
+    _processKeydownEvent(ev: KeyboardEvent): CompiledEvent {
+        this._selectAllOriginalTarget = this._getRange().startContainer;
+
+        // See comment on the same line in the _onComposition handler.
+        if (this.editable.style.display === 'none') return;
+
+        // Dead keys will trigger composition and input events later on.
+        if (ev.type === 'keydown' && ev.key === 'Dead') return;
+
+        return {
+            type: 'keydown',
+            key: ev.key,
+            altKey: ev.altKey,
+            ctrlKey: ev.ctrlKey,
+            metaKey: ev.metaKey,
+            shiftKey: ev.shiftKey,
+            defaultPrevented: ev.defaultPrevented,
+        };
+    }
+    /**
+     * @private
+     * @param {CompiledEvent} ev
+     * @param {KeyboardEvent} keypress
+     */
+    _enrichByKeypress(ev: CompiledEvent, keypress: KeyboardEvent): void {
+        if (keypress.key.length || ev.key === 'Dead') {
+            ev.key = keypress.key;
+        }
+        ev.altKey = keypress.altKey;
+        ev.ctrlKey = keypress.ctrlKey;
+        ev.metaKey = keypress.metaKey;
+        ev.shiftKey = keypress.shiftKey;
+        ev.defaultPrevented = ev.defaultPrevented || keypress.defaultPrevented;
+    }
+    /**
+     * @private
+     * @param {CompiledEvent} compiledEvent
+     * @param {InputEvent} ev
+     */
+    _enrichByInput(compiledEvent: CompiledEvent, ev: InputEvent): void {
+        compiledEvent.defaultPrevented = compiledEvent.defaultPrevented || ev.defaultPrevented;
+
+        // TODO comment: insert input data has more value than any other event
+        // type, but 'input' type has nearly no value since we don't know if it's
+        // a composition or something else, so we want to keep the original type
+        // but use the data from the input event.
+        if (!compiledEvent.type) {
+            compiledEvent.type = ev.type;
+            compiledEvent.data = ev.data;
         }
 
+        const type = compiledEvent.type;
+        const key = compiledEvent.key;
+        const data = compiledEvent.data;
+        if (type === 'composition') {
+            if (['insertCompositionText', 'insertReplacementText'].includes(ev.inputType)) {
+                // Some spell checking keyboards do not properly fill the textual
+                // data of the composition in the compositionend event. To protect
+                // against this, use the data from the input event instead.
+                compiledEvent.data = ev.data;
+                this._cloneForComposition();
```

Comment:
All these `_cloneForComposition`, aren't they too late now that we analyze the events after the tick ? Shouldn't they be in their respective `_on*` event listeners instead ?

Reply:
change proposal to fix the corrector on mac. Another solution is to give access to the vDocument in the normalizer